pkg/services/ngalert/models/instance.go

```diff
@@ -11,15 +13,121 @@ type AlertInstance struct {
 	RuleUID           string `xorm:"rule_uid"`
 	Labels            InstanceLabels
 	LabelsHash        string
-	CurrentState      InstanceStateType
+	CurrentState      InstanceState `xorm:"current_state"`
 	CurrentStateSince time.Time
 	CurrentStateEnd   time.Time
 	LastEvalTime      time.Time
 }
 
+func ErrInvalidInstanceState(i *InstanceState) error {
+	return errors.New("The InstanceState %+v is invalid")
+}
+
+// The state's Type is what the instance is doing. Alerting, Pending ...
+// The state's Reason is why the state has the Type. A common case is State
+type InstanceState struct {
+	Type   InstanceStateType  `json:"type"`
+	Reason InstanceReasonType `json:"reason"`
```

Comment:
I wonder if we can make reason optional (empty string or nil reference) and get rid of Normal.